builder/assignmentEnv.go

```diff
@@ -0,0 +1,315 @@
+package builder
+
+import (
+	"assignment-exec/image-builder/configurations"
+	"assignment-exec/image-builder/constants"
+	"assignment-exec/image-builder/environment"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/pkg/errors"
+	"io"
+	"log"
+	"os"
+	"strings"
+)
+
+type baseEnv interface {
+	verifyAndWrite() error
+	build() error
+	publish() error
+	undoWrite()
+	undoBuild() error
+	undoPublish() error
+}
+
+type assignmentEnv struct {
+	DockerfileData bytes.Buffer
+	ImgBuildConfig *imageBuildConfig
+	AssgnEnvConfig *configurations.AssignmentEnvConfig
+}
+
+type assignmentImageOption func(*assignmentEnv) error
+
+func newAssignmentImage(options ...assignmentImageOption) (*assignmentEnv, error) {
+	assgnEnv := &assignmentEnv{}
+	for _, opt := range options {
+		if err := opt(assgnEnv); err != nil {
+			return nil, errors.Wrap(err, "failed to create assignmentEnv instance")
+		}
+	}
+	return assgnEnv, nil
+}
+
+func withImageBuildCfg(imgBuildCfg *imageBuildConfig) assignmentImageOption {
+	return func(assgnEnv *assignmentEnv) error {
+		if imgBuildCfg == nil {
+			return errors.New("image build config instance not provided")
+		}
+		assgnEnv.ImgBuildConfig = imgBuildCfg
+		return nil
+
+	}
+}
+
+func withAssgnEnvConfig(assignCfgs *configurations.AssignmentEnvConfig) assignmentImageOption {
+	return func(assgnEnv *assignmentEnv) error {
+		if assignCfgs == nil {
+			return errors.New("assignment assignCfgs not provided")
+		}
+		assgnEnv.AssgnEnvConfig = assignCfgs
+		return nil
+
+	}
+}
+
+func (assgnEnv *assignmentEnv) verifyAndWrite() error {
+
+	// Verify whether language image is present in registry.
+	if err := assgnEnv.verifyLanguage(); err != nil {
+		// If no then write the instructions from base image.
+		assgnEnv.writeFromBaseImage()
+	} else {
+		if len(assgnEnv.AssgnEnvConfig.Deps.Libraries) > 0 {
+			// Else write the instructions from dependencies.
+			assgnEnv.writeFromDependencies()
+		}
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) verifyLanguage() error {
+	// Check whether the language image is available on docker hub.
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return err
+	}
+	username, found := os.LookupEnv(environment.DockerAuthUsername)
+	if !found {
+		return errors.New("environment variable for username not set")
+	}
+
+	resp, err := dockerClient.ImageSearch(backgroundContext, username, types.ImageSearchOptions{
+		Limit: 25})
+
+	if err != nil {
+		return err
+	} else {
```

Comment:
we don't need the else clause here as it is self explanatory.
```go
if err != nil {
    return err
}

// rest of the code.
```

Reply:
Changed